Add imported-from link rel to relevant SSP assemblies

aj-stein-nist opened the pull request:
# Committer Notes

{Please provide a brief description of what this PR accomplishes. Be sure to reference any issues addressed. If the PR is a work-in-progress submitted for early review, please include [WIP] at the beginning of the title or mark the PR as DRAFT.}

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/usnistgov/OSCAL/blob/main/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers
](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/OSCAL/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you updated all [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes you made? Changes to the OSCAL website can be made in the docs/content directory of your branch.
- ~Have you written new tests for your core changes, as applicable?~
- ~Have you included examples of how to use your new feature(s)?~

aj-stein-nist commented:
OK per the updated dependencies listing, I am gonna park this until #756 is implemented and circle back to this. Moving the linked issue in the TODO column.

david-waltermire-nist commented on src/metaschema/oscal_ssp_metaschema.xml:
```diff
@@ -646,6 +647,11 @@
         <enum value="no">The component does not allow an authenticated scan.</enum>
       </allowed-values>
 
+      <allowed-values target="component/link/@rel" allow-other="yes">
+        &allowed-values-component_component_link-rel;
+        <enum value="imported-from">Identifies the UUID or URI the <code>component-definition</code> that originally defined the <code>component</code>.</enum>
+      </allowed-values>
+
```
Since this targets the component directly, this constraint should be on `system-component`.

david-waltermire-nist reviewed (changes requested):
Please relocate the constraint for `component/link/@rel` to the `system-component` definition.

aj-stein-nist commented:
> Please relocate the constraint for `component/link/@rel` to the `system-component` definition.

So I pushed up a separate commit https://github.com/usnistgov/OSCAL/pull/1403/commits/497be0dc1302c567f1161770de842263e8f7e0a1 and didn't squash it yet. I understand the request, but I am not sure how this will not lead to undesirable behavior. Can you explain to me when you have a chance?

If we move it here, does this not in fact allow this prop on any any component, not _only a component_ instance in a SSP? Do we want abstracted `component-definition`(s) `component`(s) able to have this `prop` and potentially lead to confusion? Let me know what I am missing.

david-waltermire-nist commented on src/metaschema/oscal_ssp_metaschema.xml:
```diff
@@ -720,6 +721,10 @@
       <field ref="remarks" in-xml="WITH_WRAPPER"/>
     </model>
     <constraint>
+      <allowed-values target="link/@rel" allow-other="yes">
+        &allowed-values-component_component_link-rel;
+        <enum value="imported-from">Identifies the UUID or URI the <code>component-definition</code> that originally defined the <code>component</code>.</enum>
+      </allowed-values>
```
I think this may need to be removed.

david-waltermire-nist reviewed (dismissed):
LGTM. Thanks!

aj-stein-nist commented:
The failed CI job was the link checker from cache errors with the W3C website docs, it worked after re-run. @david-waltermire-nist, per discussion during the weekly status meeting, this is ready for review.

aj-stein-nist commented:
> > Please relocate the constraint for `component/link/@rel` to the `system-component` definition.
> 
> So I pushed up a separate commit [497be0d](https://github.com/usnistgov/OSCAL/commit/497be0dc1302c567f1161770de842263e8f7e0a1) and didn't squash it yet. I understand the request, but I am not sure how this will not lead to undesirable behavior. Can you explain to me when you have a chance?
> 
> If we move it here, does this not in fact allow this prop on any any component, not _only a component_ instance in a SSP? Do we want abstracted `component-definition`(s) `component`(s) able to have this `prop` and potentially lead to confusion? Let me know what I am missing.

Also just rebased down to one commit, as the feedback concern in https://github.com/usnistgov/OSCAL/pull/1403#pullrequestreview-1066766779 has been explained and addressed by Dave.

david-waltermire-nist commented:
@aj-stein-nist This PR needs to be rebased on the latest develop.

aj-stein-nist commented:
> @aj-stein-nist This PR needs to be rebased on the latest develop.

Rebased.

david-waltermire-nist commented on src/metaschema/oscal_ssp_metaschema.xml:
```diff
@@ -705,6 +706,10 @@
     </model>
     <constraint>
+      <allowed-values target="link/@rel" allow-other="yes">
+        &allowed-values-component_component_link-rel;
```
I don't think these values belong here since they are intended to be used with the link on a component.

Is there a reason they were added here?

```suggestion
```

david-waltermire-nist commented on src/metaschema/oscal_ssp_metaschema.xml:
```diff
@@ -6,6 +6,7 @@
 <!-- OSCAL Implementation Layer: System Security Plan (SSP) METASCHEMA -->
 <!-- validate with XSD and Schematron (linked) -->
 <!DOCTYPE METASCHEMA [
+  <!ENTITY allowed-values-component_component_link-rel SYSTEM "./shared-constraints/allowed-values-component_component_link-rel.ent">
```
If my comment below means this entity is not needed, then this can be removed.

```suggestion
```

david-waltermire-nist commented on src/metaschema/oscal_ssp_metaschema.xml:
```diff
@@ -705,6 +706,10 @@
     <constraint>
+      <allowed-values target="link/@rel" allow-other="yes">
+        &allowed-values-component_component_link-rel;
+        <enum value="imported-from">Identifies the UUID or URI the <code>component-definition</code> that originally defined the <code>component</code>.</enum>
```
```suggestion
        <enum value="imported-from">The hyperlink identifies a URI pointing to the <code>component</code> in a <code>component-definition</code> that originally defined the control narrative.</enum>
```

david-waltermire-nist replied:
Thinking about this more, I believe this link relation should be on the `by-component` instead, since that is where the control narrative should exist. Perhaps we should merge my suggestion above and then relocate it there?

aj-stein-nist commented on README_validations.md:
```diff
@@ -26,7 +26,7 @@ For OSCAL XML, this schema syntax is [XML Schema Definition (XSD)](https://www.w
 Each tool will have its own way to provide for validation. Bindings or configurations to link documents to schemas can be maintained either inside or outside the documents themselves; many tools offer several ways to do this.
 
-Some tools support self-contained validation functionality without full-fledged application frameworks or development environments, with tools like [`xmllint`](http://xmlsoft.org/xmllint.html) offering command-line interfaces. Similarly, you can use the [`ajv-cli`](https://github.com/ajv-validator/ajv-cli) for JSON.
+Some tools support self-contained validation functionality without full-fledged application frameworks or development environments, with tools like [`xmllint`](https://gitlab.gnome.org/GNOME/libxml2/) offering command-line interfaces. Similarly, you can use the [`ajv-cli`](https://github.com/ajv-validator/ajv-cli) for JSON.
```
Non-block comment for @david-waltermire-nist to understand what I encountered: xmlsoft.org now seems to be dead, that is what is breaking the Markdown link validation. The redirect target is bad. The HTTPS cert is not properly configured for the correct sub-domain (for *.gnome.org, not *pages.gitlab.gnome.org) and going HTTP forces HTTPS redirect to the bad target. So, I will go to the GitLab project URL, not the rendered wiki pages hosted from GitLab.

```sh
$ curl -I http://xmlsoft.org/xmllint.html 
HTTP/1.1 302 Found
Date: Thu, 25 Aug 2022 14:18:26 GMT
Server: Apache/2.2.15 (CentOS)
Location: https://gnome.pages.gitlab.gnome.org/libxml2/xmllint.html
Connection: close
Content-Type: text/html; charset=iso-8859-1

$ curl -I http://gnome.pages.gitlab.gnome.org/libxml2/xmllint.html
HTTP/1.1 302 Found
content-length: 0
location: https://gnome.pages.gitlab.gnome.org/libxml2/xmllint.html
cache-control: no-cache
```

aj-stein-nist replied:
Per weekly team sync, will review with Dave. There are concerns that GitLab landing page README doesn't really give command instructions or any helpful end user dev docs. Will work it out with Dave when we touch base.

aj-stein-nist replied:
Per conversation with @david-waltermire-nist, this seems to be a known issue and impacts all the GitLab pages hosted on gitlab.gnome.org because you cannot use LetsEncrypt (or any TLS certificate authority for that matter) to have multiple-wildcards in the SAN and [GNOME Infrastructure Team has marked this WONTFIX in their issue board](https://gitlab.gnome.org/Infrastructure/GitLab/-/issues/395). Going to use the popular die.net for its indexed man page collection, and remove the download link.

aj-stein-nist commented:
Also TIL `git commit --amend --signoff` is needed to sign-off commits that are co-authored by team members. Now the PR commits are not so unverified.

aj-stein-nist commented:
@david-waltermire-nist, I pointed this PR to a feature branch as instructed during the weekly sync for when we are ready to merge.

iMichaela commented on src/metaschema/oscal_ssp_metaschema.xml:
```diff
@@ -1020,6 +1020,10 @@
     <constraint>
+      <allowed-values target="link/@rel" allow-other="yes">
+        <!-- TODO: More work needs to be done around link enforcement constraints and requires Metaschema work on recurse-depth-first() recursive index searching. -->
+        <enum value="imported-from">The hyperlink identifies a URI pointing to the <code>component</code> in a <code>component-definition</code> that originally defined the <code>component</code>.</enum>
```
@david-waltermire-nist -- I am thinking that using "that originally **defined** the <code>component</code>" might be misleading since the CDef suggests control implementations, does not **define** them as in **deciding** them. You made recently an important correction/clarification for the CDef.

david-waltermire-nist replied:
Please propose a wording change.

david-waltermire-nist replied:
I'd like to merge this on 8/30. If @iMichaela doesn't provide a wording change suggestion, I am good with keeping this with a minor tweek.

```suggestion
        <enum value="imported-from">The hyperlink identifies a URI pointing to the <code>component</code> in a <code>component-definition</code> that originally described the <code>component</code> this component was based on.</enum>
```

aj-stein-nist replied:
Suggestion merged, will rebase and squash in a little bit.

aj-stein-nist commented:
Rebased and waiting for pending feedback, but minor recommendation in https://github.com/usnistgov/OSCAL/pull/1403#discussion_r957570701 was added before rebase. Should be good to go.

david-waltermire-nist reviewed (approved):
LGTM. Thanks!
